sympy/physics/mechanics/_system.py

```diff
@@ -0,0 +1,937 @@
+
+    @property
+    def kdes(self):
+        """Kinematic differential equations, which describe the coupling
+        between the generalized coordinates and the generalized speeds."""
+        return self._kdes
+
+    @kdes.setter
+    @_reset_eom_method
+    def kdes(self, kdes):
+        kdes = System._objects_to_list(kdes)
+        self._kdes = self._parse_expressions(
+            kdes, [], 'kinematic differential equations')
+
+    @property
+    def holonomic_constraints(self):
+        """Matrix with the holonomic constraints as rows."""
+        return self._hol_coneqs
+
+    @holonomic_constraints.setter
+    @_reset_eom_method
+    def holonomic_constraints(self, constraints):
+        constraints = System._objects_to_list(constraints)
+        self._hol_coneqs = self._parse_expressions(
+            constraints, [], 'holonomic constraints')
+
+    @property
+    def nonholonomic_constraints(self):
+        """Matrix with the nonholonomic constraints as rows."""
+        return self._nonhol_coneqs
+
+    @nonholonomic_constraints.setter
+    @_reset_eom_method
+    def nonholonomic_constraints(self, constraints):
+        constraints = System._objects_to_list(constraints)
+        self._nonhol_coneqs = self._parse_expressions(
+            constraints, [], 'nonholonomic constraints')
+
+    @property
+    def eom_method(self):
+        """Backend for forming the equations of motion."""
+        return self._eom_method
+
+    @staticmethod
+    def _objects_to_list(lst):
+        """Helper to convert passed objects to a list."""
+        if not iterable(lst):  # Only one object
+            return [lst]
+        return list(lst[:])  # converts Matrix and tuple to flattened list
+
+    @staticmethod
+    def _check_objects(objects, obj_lst, expected_type, obj_name, type_name):
+        """Helper to check the objects that are being added to the system."""
+        seen = set(obj_lst)
+        duplicates = set()
+        wrong_types = set()
+        for obj in objects:
+            if not isinstance(obj, expected_type):
+                wrong_types.add(obj)
+            if obj in seen:
+                duplicates.add(obj)
+            else:
+                seen.add(obj)
+        if wrong_types:
+            raise TypeError(f'{obj_name} {wrong_types} are not {type_name}.')
+        if duplicates:
+            raise ValueError(f'{obj_name} {duplicates} have already been added '
+                             f'to the system.')
+
+    def _parse_coordinates(self, new_coords, independent, old_coords_ind,
+                           old_coords_dep, is_coordinates=True):
+        """Helper to parse coordinates and speeds."""
+        # Construct lists of the independent and dependent coordinates
+        if is_coordinates:
+            q_ind, q_dep = old_coords_ind[:], old_coords_dep[:]
+            u_ind, u_dep = self.u_ind[:], self.u_dep[:]
+            coords_ind, coords_dep = q_ind, q_dep
```

Comment:
Why the multiple variables referencing the same thing?

Reply:
This way it is possible to use `_validate_coords`, while also using the same generic for loop. Is this playing too much with this awesome python feature, where you can kind of use pointers.

Can change it to use a more elaborate if statement for `_validate_coordinates`.

Reply:
I've looked at this again and am still not understanding what's going on here. Even if it's just me being dim, it indicates that this code might be being too clever for its own good, and therefore be less maintainable.

Reply:
Addressed in b6b695e90b1a9d6cf7b75f1f5fcfefcd9dda80a3